src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+            allocator.free(self.counts);
+            allocator.free(self.clocks);
+        }
+
+        pub fn reset(self: *Self) void {
+            mem.set(Tag, self.tags, 0);
+            mem.set(u64, self.counts, 0);
+            mem.set(u64, self.clocks, 0);
+        }
+
+        pub fn get(self: *Self, key: Key) ?*align(value_alignment) Value {
+            const set = self.associate(key);
+            return self.search(set, key);
+        }
+
+        pub const GetOrPutResult = struct {
+            value_ptr: *align(value_alignment) Value,
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
+            return self.get_or_put_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Get a pointer to an existing value with the given key if the key is in the cache.
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (Context, *align(value_alignment) const Value) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) GetOrPutResult {
+            const set = self.associate(key);
+
+            if (self.search(set, key)) |existing| {
+                assert(!locked(context, existing));
+                return .{
+                    .value_ptr = existing,
+                    .found_existing = true,
+                };
+            }
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            const max_iterations = layout.ways * math.maxInt(Count);
+            var safety_counter: usize = 0;
+
+            var way = Clocks.get(self.clocks, clock_index);
+            while (safety_counter < max_iterations + 1) : ({
+                way +%= 1;
+                safety_counter += 1;
+            }) {
+                // We pass a value pointer to the callback here so that a cache miss
+                // can be avoided if the caller is able to determine if the value is
+                // locked by comparing pointers directly.
+                if (locked(context, @alignCast(value_alignment, &set.values[way]))) continue;
+
+                const count = Counts.get(self.counts, set.offset + way);
+
+                // Free way found
+                if (count == 0) break;
+
+                Counts.set(self.counts, set.offset + way, count - 1);
+            }
+            assert(safety_counter < max_iterations);
```

Comment:
I don't know of anywhere else we start loop variables from `1` instead of `0` though, which is the only reasonable way I see to make this check use `<=` instead of `<`. I don't see a problem with using `<` here, starting loop variables/counters from `0` feels most natural and when doing so a check with `<` is 100% expected.

Reply:
> I don't know of anywhere else we start loop variables from 1 instead of 0 though

We don't typically start loop variables from 1 because we use 0-based indexing.

However, both `safety_count` and `clock_iterations_max` are not indexes but *counts*, i.e. both 1-based, hence `<=` is less surprising. Otherwise, `safety_count` should be renamed to `safety_index` but that makes less sense. It's really a count.

Reply:
That's a good point, changed it to start at 1 and use `<=`!